### PR TITLE
[Feature:Submission] Warn user when trying to upload more than the max allowed files

### LINKED ISSUE #4456: Display maximum number of files that can be uploaded to user
**What problem are you trying to solve with Submitty**
Right now across submitty's file upload drag and drop areas users can add as many files as they want even the the max_file_uploads is restricted to 20 by php (by default).

**Describe the way you'd like to solve this problem**
On every interface to upload files display the max allowed to users and then do a client side check to prevent users from adding 20+. Since php already handles checking for a number of file.

### PR DESCRIPTION
**Pull request by marwanatef2**
Fixes #4456 

Users were allowed to add any no. of files they want however, the maximum allowable no. for submission in php is restricted by default to 20.
Now users are not allowed to add more than 20 files per submission where if they tried they get alerted that they have exceeded the limit.

The new behavior can be seen in this gif
addvalid

Please let me know if any messages or something need to be changed.

### CODE REVIEW

**Comment by marwanatef2** [comment]
> the best place to actually check for the file limit is the `handleSubmission` & `handlebulk` functions

my point is i don't want user to submit 20+ files then alert some message that submission file limit is exceeded so he has to remove some files, i want to prevent him from adding more than 20 files, stopping him at the limit, so it is better handled with addition not submission.

**Comment by MasterOdin** [comment]
I agree on showing the message on adding, it's a friendlier user experience.

**Comment by bmcutler** [comment]
Instead of hardcoding "20" into our source code, can we read the current system setting?
For example, if a production installation has modified the php.ini setting to be 50 files
https://www.codexworld.com/how-to/increase-files-upload-limit-in-php/

Also, can we add a line to the warning/error message that says
"Maximum allowed number of files to be uploaded is 20.
Please package your files in a single .zip file if it is necessary for you to submit more than this limit."

**Comment by shailpatels** [comment]
@marwanatef2 are you still working on this PR?

**Comment by marwanatef2** [comment]
> @marwanatef2 are you still working on this PR?

Yes, I am. I'll get it done I was just working on my proposal.

**Comment by marwanatef2** [comment]
> Instead of hardcoding "20" into our source code, can we read the current system setting?

this part is done.

> Also, can we add a line to the warning/error message that says
> "Maximum allowed number of files to be uploaded is 20.
> Please package your files in a single .zip file if it is necessary for you to submit more than this limit."

this part can be shown in this screenshot
Capture

File: site/app/templates/submission/homework/SubmitBox.twig [hunk 1]
```diff
@@ -2,7 +2,10 @@
 
 {# Display the gradeable_message if one exists at the top of the page #}
 {% if has_gradeable_message %}
-    <div class='content gradeable_message'>{{ gradeable_message | markdown }}</div>
+    <div class='content gradeable_message'>
+        {{ gradeable_message | markdown }}
+        Maximum allowed number of files to be uploaded is {{ max_file_uploads }}.
```
Comment (MasterOdin) [inline site/app/templates/submission/homework/SubmitBox.twig:7]:
```
        {% if part_names|length > 1 %}
            <div id="upload-boxes">
                {# Submit boxes #}
                {% for part in part_names %}
                    {# uploadIndex and input-file + Index required for drag-and-drop.js #}
                    <div tabindex="0"
                        id="upload{{ loop.index }}"
                        class="upload-box"
                        onkeypress="clicked_on_box(event)"
                        role="text"
                        aria-label="Press enter to upload your {{ part }} file"
                    >
                        <h2 class="label" id="label{{ loop.index }}">
                            {% if part_names|length > 1 %}
                                Drag your {{ part }} file(s) here or click to open file browser
                            {% else %}
                                Drag your file(s) here or click to open file browser
                            {% endif %}
                        </h2>
                        <input type="file" name="files" id="input-file{{ loop.index }}" class="hide" onchange="addFilesFromInput({{ loop.index }})" multiple aria-label="Select Files to upload" />
                        <table class="file-upload-table" id="file-upload-table-{{ loop.index }}">
                        </table>
                    </div>
                {% endfor %}
                <p>Maximum allowed number of files to be uploaded is {{ max_file_uploads }}.</p>
            </div>
        {% endif %}
```

(I'm adding a surrounding `{% if %}` here as well as with your change.)

File: site/public/js/drag-and-drop.js [hunk 2]
```diff
@@ -17,6 +17,9 @@ var label_array = [];
 var use_previous = false;
 var changed = false;        // if files from previous submission changed
 
+var total_files_added = 0;
+var MAX_NO_OF_FILES;
```
Comment (MasterOdin) [inline site/public/js/drag-and-drop.js:21]: Use `MAX_NUM_OF_FILES` as the name.

Use `let` keyword please for new variables.

File: site/public/js/drag-and-drop.js [hunk 3]
```diff
@@ -219,6 +246,7 @@ function deleteFiles(part) {
     var labels = dropzone.getElementsByClassName("file-label");
     while(labels[0]){
         dropzone.removeChild(labels[0]);
+        total_files_added --;
```
Comment (MasterOdin) [inline site/public/js/drag-and-drop.js:249]: Get rid of the space here between the variable and the operator

File: site/public/js/drag-and-drop.js [hunk 4]
```diff
@@ -243,6 +271,7 @@ function deleteSingleFile(filename, part, previous) {
             if (file_array[part-1][j].name == filename) {
                 file_array[part-1].splice(j, 1);
                 label_array[part-1].splice(j, 1);
+                total_files_added--;
```
Comment (MasterOdin) [inline site/public/js/drag-and-drop.js:274]: Get rid of the space here between the variable and the operator

**Comment by marwanatef2** [comment]
> your message is causing the submission area to be pushed to the left side

it looks like I added the message outside the "upload box" div, modified now.
 
> On chrome after I submit past the max number of files, and get the alert, and I try to submit more than the max number again I don't get the alert, nothing seems to happen. This appears to be on chrome only, firefox does not have this issue on Mac.

I am working on chrome already and I had the same problem, It's fixed now as shown in the gif below.

readding

**Comment by bmcutler** [comment]
@shailpatels @MasterOdin 
Can one or both of you please re-review & approve

**Pull request merged**